### PR TITLE
replace `flake8` with `ruff`

### PR DESCRIPTION
**Pull request by zacharyburnett**
`ruff` is an in-development Python linter written in Rust that aims for rule-checking parity with Flake8 and Black: https://github.com/charliermarsh/ruff

It has several advantages over `flake8`, including significant speedups in linting and support for `pyproject.toml` configuration and PEP517 / PEP621.

**Checklist**
- [x] added entry in `CHANGES.rst` (either in `Bug Fixes` or `Changes to API`)
- [ ] updated relevant tests
- [ ] updated relevant documentation
- [ ] updated relevant milestone(s)
- [x] added relevant label(s)

### CODE REVIEW

**Review by mwregan2 — APPROVED** [review]

File: pyproject.toml [hunk 1]
```diff
@@ -65,3 +65,16 @@ doctest_rst = true
...
+ignore = [
+    'C901', # variable is too complex
+]
+exclude = [
+    'docs',
+    'build',
+    'dist',
+    '.tox',
+    '.eggs',
+]
```
Comment (hbushouse) [inline pyproject.toml:80]: Looks like a minor formatting issue here?

**Review by hbushouse — APPROVED** [review]

File: pyproject.toml (hunk 1 above)
Comment (zacharyburnett) [reply pyproject.toml:80]: `toml` allows a trailing comma, but I can remove it if you prefer. Is that what you were referencing?

Comment (hbushouse) [reply pyproject.toml:80]: No, referring to the popup regarding no newline *after* line 80.

Comment (zacharyburnett) [reply pyproject.toml]: oh, I didn't even notice that 😅

**Pull request merged**